### ISSUE TITLE
Export silero-vad ONNX to TensorRT

### ISSUE THREAD

**Issue by EnlNovius** [snakers4/silero-vad#208]
## Export silero-vad ONNX to TensorRT

I am trying to translate the supplied ONXX network (`files/silero_vad.onnx`) to TensorRT (.trt). I tried two tools:
- `trtexec` from nvidia (https://github.com/NVIDIA/TensorRT/tree/main/samples/trtexec)
- `onnx2trté` (https://github.com/onnx/onnx-tensorrt)

With both tools, I get the following error:
```shell
$ trtexec --onnx=files/silero_vad.onnx
...
[08/03/2022-14:44:43] [E] Error[4]: [shuffleNode.cpp::symbolicExecute::392] Error Code 4: Internal Error (Reshape_17: IShuffleLayer applied to shape tensor must have 0 or 1 reshape dimensions: dimensions were [-1,2])
[08/03/2022-14:44:43] [E] [TRT] ModelImporter.cpp:773: While parsing node number 27 [Pad -> "152"]:
[08/03/2022-14:44:43] [E] [TRT] ModelImporter.cpp:774: --- Begin node ---
[08/03/2022-14:44:43] [E] [TRT] ModelImporter.cpp:775: input: "129"
...
[08/03/2022-14:44:43] [E] [TRT] ModelImporter.cpp:776: --- End node ---
[08/03/2022-14:44:43] [E] [TRT] ModelImporter.cpp:779: ERROR: ModelImporter.cpp:180 In function parseGraph:
[6] Invalid Node - Pad_27
[shuffleNode.cpp::symbolicExecute::392] Error Code 4: Internal Error (Reshape_17: IShuffleLayer applied to shape tensor must have 0 or 1 reshape dimensions: dimensions were [-1,2])
[08/03/2022-14:44:43] [E] Failed to parse onnx file
```

From what I could find, the problem would come from the fact that TensorRt does not currently manage tensor2D (https://forums.developer.nvidia.com/t/ishufflelayer-applied-to-shape-tensor-must-have-0-or-1-reshape-dimensions-dimensions-were-1-2/200183). A solution proposed in response is to use `polygraphy` with surgeon:

```
$  polygraphy surgeon sanitize --fold-constants files/silero_vad.onnx -o files/silero_vad_surgeon.onnx
```

Now if I apply one of my two tools, the problem seems to be solved, but another problem arises further on:

```
$ trtexec --onnx=files/silero_vad_surgeon.onnx
...
[08/03/2022-14:46:22] [E] Error[4]: If_33_OutputLayer: IIfConditionalOutputLayer inputs must have the same shape.
[08/03/2022-14:46:22] [E] [TRT] ModelImporter.cpp:773: While parsing node number 10 [If -> "158"]:
[08/03/2022-14:46:22] [E] [TRT] ModelImporter.cpp:774: --- Begin node ---
[08/03/2022-14:46:22] [E] [TRT] ModelImporter.cpp:775: input: "157"
...
[08/03/2022-14:46:22] [E] [TRT] ModelImporter.cpp:776: --- End node ---
[08/03/2022-14:46:22] [E] [TRT] ModelImporter.cpp:779: ERROR: ModelImporter.cpp:180 In function parseGraph:
[6] Invalid Node - If_33
If_33_OutputLayer: IIfConditionalOutputLayer inputs must have the same shape.
[08/03/2022-14:46:22] [E] Failed to parse onnx file
```

I haven't found a solution to this problem yet, does anyone have any idea how to solve this problem?

**Comment by snakers4** [comment]
> TensorRt does not currently manage tensor2D

Interesting, 95% of the network consists of either 1D convolutions or Linear layers. 
The network contains an internal normalization layer (with a padding), which most likely is the cause of the problem, since it always gave us grief during exports.
This is one of the reasons we decided to keep PyTorch and ONNX formats.

> I am trying to translate the supplied ONXX network (files/silero_vad.onnx) to TensorRT (.trt).

But the main question is, why?

**Comment by EnlNovius** [comment]
> But the main question is, why?

I'm trying to use silero-vad in real time on an embedded system already using several tensorrt neural networks. Switching from a pytorch model to a tensorrt model normally allows optimization at the inference level (https://developer.nvidia.com/blog/speeding-up-deep-learning-inference-using-tensorflow-onnx-and-tensorrt/).

**Comment by EnlNovius** [comment]
I just tried onnx_simplifier (https://github.com/daquexian/onnx-simplifier). The simplification of the tree solves both the first and the second problem. 

```bash
$ onnxsim files/silero_vad.onnx files/silero_vad_onnxsim.onnx
```

However, a new problem arises later on:

```shell
$ trtexec --onnx=files/silero_vad_onnxsim.onnx
...
08/03/2022-16:00:23] [E] Error[4]: [graphShapeAnalyzer.cpp::processCheck::587] Error Code 4: Internal Error (Conv_81: spatial dimension of convolution output cannot be negative (build-time output dimension of axis 2 is -5))
[08/03/2022-16:00:23] [E] Error[2]: [builder.cpp::buildSerializedNetwork::636] Error Code 2: Internal Error (Assertion engine != nullptr failed. )
...
```

**Comment by snakers4** [comment]
> on a 4 minutes audio file

One audio chunk [should take](https://github.com/snakers4/silero-vad/wiki/Performance-Metrics) about ~1ms on one CPU thread.
ONNX was similar.

By a simple calculation, a 4 minute audio file should take about:

- 4 mins * 60 seconds per minute * 16000 samples per second / 1536 samples per chunk * 1 ms per chunk / 1000 ms per second = 2.5 seconds

The fact that this takes ~40 times slower is strange.

> import pycuda.autoinit # You need this to init cuda

There is very little reason to run VAD on CUDA, because there is very little to gain.
The model is very fast, most likely it will just incur an overhead to copy to and from the GPU.

**Comment by EnlNovius** [comment]
> The fact that this takes ~40 times slower is strange.

I wasn't clear in my results, I edited the post to correct that. The results that were displayed were those for running the network 100 times over 4min of video, so we were at 0.5ms / frame. 

> There is very little reason to run VAD on CUDA, because there is very little to gain.
> The model is very fast, most likely it will just incur an overhead to copy to and from the GPU.

For my project, the audio data is already on the gpu, since it goes through other networks. According to the documentation *"Using batching or GPU can also improve performance considerably.*. Having already the data on GPU and the GPU being supposed to accelerate the processing times a bit more, I wanted to try to see what it could give, even if the performance on silero CPU is already remarkable. 

Since I managed to run silero on TensorRT, I think the issue is complete. Thanks for the answers and for the very good work done on silero-vad :smiley:

**Comment by snakers4** [comment]
> so we were at 0.5ms / frame.

Seems in line with our bechmarks, albeit we did not test on GPU.

> "Using batching or GPU can also improve performance considerably.. Having already the data on GPU and the GPU being supposed to accelerate the processing times a bit more, 

Well, batching would work for multiple [streams](https://github.com/snakers4/silero-vad/discussions/142) at the same time better. You can find more details in the discussion via this link. Basically each batch element in one separate stream.

In any case many thanks for your input on this conversion.
Hope someone finds it useful for their usecase.

> Since I managed to run silero on TensorRT

Another questions is whether the model outputs are similar.

I will create a copy of this ticket as a discussion.
